src/x509.rs

```diff
@@ -157,6 +157,42 @@ impl<S: BERDecodable + Integer, A: BERDecodable + SignatureAlgorithm, K: BERDeco
     }
 }
 
+/// BasicConstraints type as defined in RFC 5280.
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct BasicConstraints {
+    pub ca: bool,
+    pub path_len_constraint: Option<u64>,
+}
+
+impl HasOid for BasicConstraints {
+    fn oid() -> &'static ObjectIdentifier {
+        &oid::basicConstraints
+    }
+}
+
+impl DerWrite for BasicConstraints {
+    fn write(&self, writer: DERWriter) {
+        writer.write_sequence(|writer| {
+            if self.ca { // do not write field if `ca` is equal to default value `false`
+                writer.next().write_bool(true);
+            }
+            if let Some(path_len) = self.path_len_constraint {
```

Comment:
I may be too tired to review, but the RFC states:
> The pathLenConstraint field is meaningful only if the cA boolean is
> asserted and the key usage extension, if present, asserts the
> keyCertSign bit.
So should't it be within the `if self.ca`  then case?

Reply:
I'd argue that this specifies how the `BasicConstraints` type should be *interpreted*, not how it should be encoded.

The RFC also states:

> CAs MUST NOT include the pathLenConstraint field unless the cA
   boolean is asserted and the key usage extension asserts the
   keyCertSign bit.

But again, I would argue that this specifies how a CA should *use* the `BasicConstraints` type, not how it should be encoded.

If we'd really want to enforce use of this type according to these statements from the RFC, then I think the rust type should be defined as:
```
pub enum BasicConstraints {
    Ca(Option<u64>),
    NoCa,
}
```
That would statically ensure that the `pathLenConstraint` field can only be specified when the `cA` field is `true`.

But I think this crate should not be that opinionated, and instead leave it up to its users to determine how to properly use and interpret the `BasicConstraints` type.

Reply:
After offline discussion: will update PR to use the `BasicConstraints` enum as defined above.